File: 2023/day5.py
from functools import reduce
import re

def load():

    with open("_input/day5.txt", encoding='utf8') as f:
        lines = f.read().splitlines()

    seeds = list(map(lambda x: int(x), lines[0].split('seeds: ')[1].split(' ')))
    maps = {}
    load = ''
    for line in lines:
        if line == 'seed-to-soil map:':
            load = 0 #'seed-to-soil'
            maps[load] = []
            continue
        if line == 'soil-to-fertilizer map:':
            load = 1 #'soil-to-fertilizer'
            maps[load] = []
            continue
        if line == 'fertilizer-to-water map:':
            load = 2 #'fertilizer-to-water'
            maps[load] = []
            continue
        if line == 'water-to-light map:':
            load = 3 #'water-to-light'
            maps[load] = []
            continue
        if line == 'light-to-temperature map:':
            load = 4 #'light-to-temperature'
            maps[load] = []
            continue
        if line == 'temperature-to-humidity map:':
            load = 5 #'temperature-to-humidity'
            maps[load] = []
            continue
        if line == 'humidity-to-location map:':
            load = 6 #'humidity-to-location'
            maps[load] = []
            continue
        if load != '' and line != '':
            maps[load].append(list(map(lambda x: int(x), line.split(' '))))
    return seeds, maps

seeds, maps = load()

def map_value(seed, maps):
    for map in maps:
        if seed >= map[1] and seed <= map[1] + map[2]:
            return seed + (map[0] - map[1])
    return seed

def rev_map_value(seed, maps):
    for map in reversed(maps):
        if seed >= map[0] and seed <= map[0] + map[2]:
            return seed + (map[1] - map[0])
    return seed

def do_map(seed):
    m = seed
    for i in range(7):
        m = map_value(m, maps[i])
    return m

def in_seeds(seeds, n):
    for i in range(0, len(seeds), 2):
        if seeds[i] <= n and n < seeds[i] + seeds[i+1]:
            return True
    return False

def part1(seeds):
    f = []
    for seed in seeds:
        f.append(do_map(seed))
    return min(f)
    

def rev_do_map(seeds, location):
    m = location
    for i in range(6, -1, -1):
        m = rev_map_value(m, maps[i])
    if in_seeds(seeds, m):
        return m
    return False

def part2(seeds, block=1000):
    n = 0
    while rev_do_map(seeds, n) == False:
        n += block
    low = n
    for i in range(n, n-block, -1):
        r = rev_do_map(seeds, i)
        if r != False and low > i:
            low = i
    return low

print(f'part1: {part1(seeds)} part2: {part2(seeds)}')
print('it wants part 2 to be one less')

File: 2023/day5.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func load() ([]int, map[int][][]int) {
	content, err := os.ReadFile("_input/day5.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(content), "\n")
	seedsRaw := strings.Split(lines[0], "seeds: ")[1]
	seedsSplit := strings.Split(seedsRaw, " ")

	var seeds []int
	for _, seed := range seedsSplit {
		val, _ := strconv.Atoi(seed)
		seeds = append(seeds, val)
	}

	maps := make(map[int][][]int)
	load := -1

	for _, line := range lines {
		switch line {
		case "seed-to-soil map:", "soil-to-fertilizer map:", "fertilizer-to-water map:",
			"water-to-light map:", "light-to-temperature map:", "temperature-to-humidity map:",
			"humidity-to-location map:":
			load++
			maps[load] = [][]int{}
		default:
			if load != -1 && line != "" {
				splitLine := strings.Split(line, " ")
				var mapValues []int
				for _, val := range splitLine {
					mapVal, _ := strconv.Atoi(val)
					mapValues = append(mapValues, mapVal)
				}
				maps[load] = append(maps[load], mapValues)
			}
		}
	}

	return seeds, maps
}

func mapValue(seed int, maps [][]int) int {
	for _, m := range maps {
		if seed >= m[1] && seed <= m[1]+m[2] {
			return seed + (m[0] - m[1])
		}
	}
	return seed
}

func revMapValue(seed int, maps [][]int) int {
	for i := len(maps) - 1; i >= 0; i-- {
		m := maps[i]
		if seed >= m[0] && seed <= m[0]+m[2] {
			return seed + (m[1] - m[0])
		}
	}
	return seed
}

func doMap(seed int, maps map[int][][]int) int {
	m := seed
	for i := 0; i < 7; i++ {
		m = mapValue(m, maps[i])
	}
	return m
}

func inSeeds(seeds []int, n int) bool {
	for i := 0; i < len(seeds); i += 2 {
		if seeds[i] <= n && n < seeds[i]+seeds[i+1] {
			return true
		}
	}
	return falseHi
}

func part1(seeds []int, maps map[int][][]int) int {
	f := []int{}
	for _, seed := range seeds {
		f = append(f, doMap(seed, maps))
	}
	minimum := f[0]
	for _, val := range f {
		if val < minimum {
			minimum = val
		}
	}
	return minimum
}

func revDoMap(seeds []int, location int, maps map[int][][]int) int {
	m := location
	for i := 6; i >= 0; i-- {
		m = revMapValue(m, maps[i])
	}
	if inSeeds(seeds, m) {
		return m
	}
	return -1
}

func part2(seeds []int, maps map[int][][]int, block int) int {
	n := 0
	for revDoMap(seeds, n, maps) == -1 {
		n += block
	}
	low := n
	for i := n; i > n-block; i-- {
		r := revDoMap(seeds, i, maps)
		if r != -1 && low > i {
			low = i
		}
	}
	return low
}

func main() {
	seeds, maps := load()

	resultPart1 := part1(seeds, maps)
	resultPart2 := part2(seeds, maps, 1000)

	fmt.Printf("part1: %d part2: %d\n", resultPart1, resultPart2-1)
	fmt.Println("it wants part 2 to be one less")
}
